crates/blockchain-tree/src/shareable.rs

```diff
@@ -108,6 +108,60 @@ impl<DB: Database, C: Consensus, EF: ExecutorFactory> BlockchainTreeViewer
         let (number, blocks) = self.tree.read().block_indices().pending_blocks();
         blocks.first().map(|&hash| BlockNumHash { number, hash })
     }
+
+    // TODO: better errors for the unexpected cases here (unknown fork block with index etc.)
+    fn share_chain(&self, first: &BlockHash, second: &BlockHash) -> Result<bool, Error> {
+        trace!(target: "blockchain_tree", ?first, ?second, "Returning whether or not the two blocks share a chain");
+        let tree = self.tree.read();
+        let indices = tree.block_indices();
+        let (non_canon_index, maybe_canon_hash) =
+            match (indices.get_blocks_chain_id(first), indices.get_blocks_chain_id(second)) {
+                // both part of side chains in the tree
+                (Some(first_index), Some(second_index)) => return Ok(first_index == second_index),
+                (Some(first_index), None) => {
+                    // if the first is not canonical, and returned None for chain id, exit early
+                    if !tree.ensure_block_is_canonical(second)? {
+                        return Ok(false)
+                    }
+                    (first_index, second)
+                }
+                (None, Some(second_index)) => {
+                    // if the first is not canonical, and returned None for chain id, exit early
+                    if !tree.ensure_block_is_canonical(first)? {
+                        return Ok(false)
+                    }
+                    (second_index, first)
+                }
+                (None, None) => {
+                    // both part of canonical chain (or do not exist), exit
+                    return Ok(tree.ensure_block_is_canonical(first)? &&
```

Comment:
not familiar with the tree internals so no idea what's going on here, we definitely need some tests for this

Reply:
yep, will add tests